### DISCUSSION TITLE
Better hot code reloading

### DISCUSSION THREAD

**Discussion by Zerotask** [vlang/v#11826] [Ideas]
It would be great if [hot code reloading](https://github.com/vlang/v/blob/master/doc/docs.md#hot-code-reloading) / watching works without using the `[live]` attribute. This is especially useful when working with vweb or in a web context in general.

For example in the NodeJS world you can use packages like [nodemon](https://nodemon.io/) (e.g. `nodemon main.js`) and further changes to this file are automatically detected and it gets compiled.

When working with V/vweb I'd like to use for example `v run -live main.v` and after it I can focus on developing without stopping and starting the web server manually every time I change something.

**Comment by dumblob** [comment]
Or alternatively simply leverage the speedy compilation of *V* and add a new `v mk` command working in a similar fashion as [Ekam](https://github.com/capnproto/ekam ) (basically just an infinite loop [watching any changes](https://github.com/facebook/watchman ) to any file or directory recursively and if change detected, then stop the running (sub)project, recompile it immediately in parallel and run it again).

No need for hot code reloading (which is anyway a bit flaky feature considering what *V* supports and aims for).

**Comment by dumblob** [comment]
Now *V* has Ekam-like fully automated recompilation: https://github.com/vlang/v/pull/9577 . Thanks @spytheman !

**Comment by fnetX** [comment]
@spytheman thank you for the mentioned PR. It looks like you are currently watching for all file changes in the repo? Do you mind watching only .v (and maybe .html template) files? Or some different pattern, like .v, and everything in templates (currently hardcoded for vweb AFAIK)? Or even customizable?

My problem is the following: I'm using sqlite and my app restarts with every request because the write apparently triggers the app to recompile :see_no_evil: - which is kind of a breaker since it introduces a race condition. But the feature itself is super useful! I am using `v -watch run .`

Best thing would of course be to watch all files opened at build time (to detect embed_files etc calls). But this is probably a bit tricky :grimacing:

**Comment by spytheman** [comment]
> @spytheman thank you for the mentioned PR. It looks like you are currently watching for all file changes in the repo? Do you mind watching only .v (and maybe .html template) files? Or some different pattern, like .v, and everything in templates (currently hardcoded for vweb AFAIK)? Or even customizable?
> 
> My problem is the following: I'm using sqlite and my app restarts with every request because the write apparently triggers the app to recompile :see_no_evil: - which is kind of a breaker since it introduces a race condition. But the feature itself is super useful! I am using `v -watch run .`
> 
> Best thing would of course be to watch all files opened at build time (to detect embed_files etc calls). But this is probably a bit tricky :grimacing: 


Try putting the sqlite db in its own subfolder. -watch works by monitoring all files in folders, that contained .v files during the initial compilation => if you put the db in a different folder/subfolder, it will be skipped.

**Comment by spytheman** [comment]
I'll think about adding a customisable way to blacklist specific file extensions/patterns so that changes to them will be ignored in -watch mode.

**Comment by dumblob** [comment]
> I'll think about adding a customisable way to blacklist specific file extensions/patterns so that changes to them will be ignored in -watch mode.

Instead of extensions let me recommend basing it on full/relative path basis while supporting "standard" widely known wildcards (which were originally designed exactly for this quick file-matching purpose).

Actually Ekam (as mentioned above https://github.com/vlang/v/issues/9154#issuecomment-792898383 ) does something similar but by having a config file in each directory (which I personally find inefficient, but it's easier to implement and use than implementing & using recursion in one config_file/command_line_arg).
